pingrid/impl.py

```diff
@@ -90,6 +79,70 @@ def error_fig(error_msg="error"):
 FuncInterp2d = Callable[[Iterable[np.ndarray]], np.ndarray]
 
 
+class ColorScale:
+    
+    def __init__(self, name, colors, scale=None, ncolors=256):
+        self.name = name
+        self.colors = colors
+        self.ncolors = ncolors
+        if scale is None:
+            self.scale = list(np.arange(len(colors)))
+        else:
+           if (np.diff(scale) < 0).any():
+              raise Exception("scale must be monotically increasing")
+           elif len(colors) == len(scale):
+               self.scale = scale
+           else:
+               raise Exception("scale must be same length as colors")
+
+    def get_name(self):
+        return self.name
+    
+    def reversed(self, name=None):
+        if name is None:
+            name = self.name + "_r"
+        return ColorScale(name, self.colors[::-1], self.scale, self.ncolors)
+
+    def rescaled(self, new_min, new_max):
+        cs_val = np.array(self.scale)
+        scale = (cs_val - cs_val[0]) * (new_max - new_min) / (cs_val[-1] - cs_val[0]) + new_min
+        return ColorScale(self.name, self.colors, scale, self.ncolors)
+
+    def to_rgba_array(self, lutsize=None):
+        if lutsize is None:
+            lutsize = self.ncolors
+        cs_i = self.rescaled(0, lutsize-1).scale.astype(int)
+        cs_i = cs_i + np.append([0], np.where(np.diff(cs_i) == 0, 1, 0))
+        cs_rgba = np.array(self.colors)
+        cs_rgba_full = np.full((lutsize, 4), np.nan)
+        for rgba in range(4):
+            cs_rgba_full[:, rgba] = np.interp(np.arange(lutsize), cs_i, cs_rgba[:, rgba])
+        return cs_rgba_full.astype(int)
+
+    def to_bgra_array(self, lutsize=None):
+        return self.to_rgba_array(lutsize=lutsize)[:,[2, 1, 0, 3]]
+
+    def to_dash_leaflet(self, lutsize=None):
+        if lutsize is None:
+            lutsize = self.ncolors
+        cm = self.to_rgba_array(lutsize=lutsize)
+        cs = []
+        for x in cm:
+            v = RGBA(*x)
+            cs.append(v.to_hex())
+        return cs
+
+
+class RGBA(NamedTuple):
+    red: int
+    green: int
+    blue: int
+    alpha: int = 255
+
+    def to_hex(self):
+        return f"#{self.red:02x}{self.green:02x}{self.blue:02x}{self.alpha:02x}"
+
+
 class BGRA(NamedTuple):
```

Comment:
I don't think we should have both RGBA and BGRA classes. We could have one class called Color, with methods like `to_hex_rgba` and `to_hex_bgra`, if needed.

Reply:
Right now there is still RGBA in pingrid and directly in fbfmaproom. I'll clean that up when we are happy with what we are doing and it all works with the enactsmulti maprooms. 

So what would the Color class look like?

Reply:
Just like RGBA, except with `to_hex` renamed `to_hex_rgba`, and a new method `to_hex_bgra` added. When creating a Color from BGRA data, we'll have to be careful either to invert the order of the color arguments, or use kwarg syntax e.g. `Color(blue=b, green=g, red=r, alpha=a)`.